Rules/ReviewUnusedParameter.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation.Language;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+#if !CORECLR
+using System.ComponentModel.Composition;
+#endif
+using System.Globalization;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// ReviewUnusedParameter: Check that all declared parameters are used in the script body.
+    /// </summary>
+#if !CORECLR
+[Export(typeof(IScriptRule))]
+#endif
+    public class ReviewUnusedParameter : IScriptRule
+    {
+        public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null)
+            {
+                throw new ArgumentNullException(Strings.NullAstErrorMessage);
+            }
+
+            var scriptBlockAsts = ast.FindAll(oneAst => oneAst is ScriptBlockAst, true);
+            if (scriptBlockAsts == null)
+            {
+                yield break;
+            }
+
+            foreach (ScriptBlockAst scriptBlockAst in scriptBlockAsts)
+            {
+                // find all declared parameters
+                IEnumerable<Ast> parameterAsts = scriptBlockAst.FindAll(oneAst => oneAst is ParameterAst, false);
+
+                // list all variables
+                IEnumerable<string> variables = scriptBlockAst.FindAll(oneAst => oneAst is VariableExpressionAst, false)
+                    .Cast<VariableExpressionAst>()
+                    .Select(variableExpressionAst => variableExpressionAst.VariablePath.ToString());
+
+                foreach (ParameterAst parameterAst in parameterAsts)
+                {
+                    // compare the list of variables to the parameter name
+                    // there should be at least two matches of the variable name since the parameter declaration counts as one
+                    int matchCount = variables

```

Comment:
At this point, the algorithm becomes O(n^2) in the number of parameters, which is undesirable.

Instead, it should be possible to do the following:

1. Create an empty case insensitive hashset of variable names
1. Collect all the variable names used in the begin, process and end blocks of the script block
1. Run through the parameters and warn for each one not found in the hashset

Reply:
@rjmholt Forgive me, but I'm failing to see how that will reduce the complexity. Isn't the current code doing essentially what you've described?

Reply:
It *almost* does, but currently:

- The variable lookup is O(n) in the number of variables. The more variables that occur in a script, the longer it will take to verify that parameter is used in that script. Worse, we must look through the whole list of variables for each parameter. With particularly large scripts, the latency of that will start to hit (visible particularly in the VSCode scenario). Instead, we should use an O(1) lookup data structure like a `HashSet<string>`
- Variable name comparison is currently case sensitive (done with `.Where(variable => variable == parameterAst.Name.VariablePath.ToString())`). Instead it should be case insensitive. This is easy to do with `new HashSet<string>(StringComparer.OrdinalIgnoreCase)`.

Reply:
Yeah, it's definitely case-sensitive at the moment. I'll take a look at implementing the hashset - I think I can some up with something. Thanks!